Crypto/ECC/Ed25519.hs

```diff
@@ -0,0 +1,298 @@
+import           Foreign.Storable
+
+import           Crypto.Error
+import           Crypto.Internal.ByteArray (ByteArrayAccess, Bytes,
+                                            ScrubbedBytes, withByteArray)
+import qualified Crypto.Internal.ByteArray as B
+import           Crypto.Internal.Compat
+import           Crypto.Internal.Imports
+import           Crypto.Random
+
+
+scalarArraySize :: Int
+scalarArraySize = 40 -- maximum [9 * 4 {- 32 bits -}, 5 * 8 {- 64 bits -}]
+
+-- | A scalar modulo order of curve Ed25519.
+newtype Scalar = Scalar ScrubbedBytes
+    deriving (Show,NFData)
+
+instance Eq Scalar where
+    (Scalar s1) == (Scalar s2) = unsafeDoIO $
+        withByteArray s1 $ \ps1 ->
+        withByteArray s2 $ \ps2 ->
+            fmap (/= 0) (ed25519_scalar_eq ps1 ps2)
+    {-# NOINLINE (==) #-}
+
+pointArraySize :: Int
+pointArraySize = 160 -- maximum [4 * 10 * 4 {- 32 bits -}, 4 * 5 * 8 {- 64 bits -}]
+
+-- | A point on curve Ed25519.
+newtype Point = Point Bytes
+    deriving NFData
+
+instance Show Point where
+    showsPrec d p =
+        let bs = pointEncode p :: Bytes
+         in showParen (d > 10) $ showString "Point "
+                               . shows (B.convertToBase B.Base16 bs :: Bytes)
+
+instance Eq Point where
+    (Point p1) == (Point p2) = unsafeDoIO $
+        withByteArray p1 $ \pp1 ->
+        withByteArray p2 $ \pp2 ->
+            fmap (/= 0) (ed25519_point_eq pp1 pp2)
+    {-# NOINLINE (==) #-}
+
+-- | Generate a random scalar.
+scalarGenerate :: MonadRandom randomly => randomly Scalar
+scalarGenerate = unwrap . scalarDecodeLong . clamp <$> generate
+  where
+    unwrap (CryptoPassed x) = x
+    unwrap (CryptoFailed _) = error "scalarGenerate: assumption failed"
+
+    generate :: MonadRandom randomly => randomly ScrubbedBytes
+    generate = getRandomBytes 32
+
+    -- Uses the same bit mask than during key-generation procedure,
+    -- but without making divisible by 8.  As a consequence of modular
+    -- reduction, distribution is not uniform.  But the curve order is
+    -- very close to 2^252 so only a tiny fraction of the scalars have
+    -- lower probability, roughly 1/(2^126) of all possible values.
+    clamp :: ByteArrayAccess ba => ba -> ScrubbedBytes
+    clamp bs = B.copyAndFreeze bs $ \p -> do
+                   b31 <- peekElemOff p 31 :: IO Word8
+                   pokeElemOff p 31 ((b31 .&. 0x7F) .|. 0x40)
+
+-- | Serialize a scalar to binary, i.e. a 32-byte little-endian
+-- number.
+--
+-- Format is binary compatible with 'Crypto.PubKey.Curve25519.SecretKey'
+-- from module "Crypto.PubKey.Curve25519".
+scalarEncode :: B.ByteArray bs => Scalar -> bs
+scalarEncode (Scalar s) =
+    B.allocAndFreeze 32 $ \out ->
+        withByteArray s $ \ps -> ed25519_scalar_encode out ps
+
+-- | Deserialize a little-endian number as a scalar.  Input array can
+-- have any length from 0 to 64 bytes.
+scalarDecodeLong :: B.ByteArrayAccess bs => bs -> CryptoFailable Scalar
```

Comment:
I'm not sure about this. The way I see it that the extended secret key of ed25519 is actually a two part type; one half is a 32 bytes scalar (in little endian) that is technically the secret value which derive the public key. While the second half is a prefix used during the signature process only.

at the ed25519 level, the 32 bytes secret key is hashed with SHA512 to make this 64 bytes extended quantity; I wonder if that would be a good idea to distinguish this here at the type level instead of considering it an opaque scalar.

Reply:
Yes ideally decoding should accept a scalar of any length and never fail. In libdecaf this is like this. In ed25519-donna, the code uses `unsigned char work[64]` as temporary buffer, so here I'm just exposing and documenting the limitation in Haskell as well.